### ISSUE TITLE
bug(sim): Twitter-Recommender rankt über zufällig initialisierte Pooler-Gewichte — twhin-bert ist ein MLM-Checkpoint ohne Pooler

### ISSUE THREAD

**Issue by arn0ld87** [arn0ld87/agora#1236]
## Symptom

Bei jedem Simulationsstart meldet der OASIS-Subprozess beim Laden von `Twitter/twhin-bert-base`:

```
[transformers] BertModel LOAD REPORT from: Twitter/twhin-bert-base
Key                                                                  | Status     |
---------------------------------------------------------------------+------------+-
bert.encoder.layer.{0...11}.attention.self.distance_embedding.weight | UNEXPECTED |
cls.predictions.transform.dense.weight                               | UNEXPECTED |
cls.predictions.transform.LayerNorm.bias                             | UNEXPECTED |
cls.predictions.transform.LayerNorm.weight                           | UNEXPECTED |
cls.predictions.transform.dense.bias                                 | UNEXPECTED |
cls.predictions.bias                                                 | UNEXPECTED |
pooler.dense.bias                                                    | MISSING    |
pooler.dense.weight                                                  | MISSING    |
```

`MISSING` heisst laut Transformers-Report: neu initialisiert, also zufällige Gewichte.

## Warum das zählt

Der Twitter-Recommender benutzt genau diese Schicht. Die Kette ist vollständig belegbar:

```
DefaultPlatformType.TWITTER  →  recsys_type="twhin-bert"      oasis/environment/env.py:81
                             →  RecsysType.TWHIN               oasis/social_platform/platform.py:343
                             →  rec_sys_personalized_twh
                             →  process_batch
                             →  return outputs.pooler_output   oasis/social_platform/process_recsys_posts.py:33
```

`DefaultPlatformType.REDDIT` setzt `recsys_type="reddit"` (`env.py:92`) und läuft über `rec_sys_reddit` ohne BERT. **Nur die Twitter-Seite ist betroffen.**

### Ursache

`Twitter/twhin-bert-base` ist ein **Masked-LM-Checkpoint** — daher die `cls.predictions.*`-Keys als UNEXPECTED. Ein MLM-Checkpoint enthält keine Pooler-Gewichte. `AutoModel.from_pretrained` baut daraus trotzdem ein `BertModel` **mit** Pooler und initialisiert `pooler.dense.weight` / `pooler.dense.bias` zufällig.

Der Recommender liest dann `pooler_output`, also `tanh(W · h_CLS + b)` mit zufälligem `W` und `b`.

### Zwei Folgen

**1. Die Post-Vektoren tragen kaum Semantik.** Eine Zufallsprojektion erhält zwar einen Teil der Geometrie, aber der untrainierte `tanh` sättigt, und der CLS-Token eines BERT ohne trainierten Pooler ist ohnehin kein brauchbares Satz-Embedding. Was ein Agent auf Twitter im Feed sieht, ist damit weitgehend beliebig — und der Feed bestimmt, worauf er reagiert.

**2. Es ist nicht reproduzierbar.** Ein `torch.manual_seed` findet sich weder in `backend/scripts/` noch in `backend/app/` noch im OASIS-Aufrufpfad. Die Zufallsgewichte unterscheiden sich also pro Prozessstart. Zwei Läufe auf identischen Daten bekommen unterschiedliche Twitter-Rankings.

Punkt 2 ist der teurere: Er macht Läufe auf der Twitter-Seite grundsätzlich nicht vergleichbar. Jede Aussage der Form „in Lauf A verhielt sich Twitter anders als in Lauf B" trägt einen unbekannten Anteil Initialisierungsrauschen. Für ein Werkzeug, dessen Zweck der Vergleich simulierter Reaktionen ist, ist das eine stille Untergrabung der Aussagekraft.

## Vorschlag

**Mean-Pooling über `last_hidden_state` mit Attention-Maske** statt `pooler_output`. Das ist der Standardweg für Satz-Embeddings aus einem Encoder ohne trainierten Pooler und braucht keine zusätzlichen Gewichte:

```python
out = model(**inputs)
mask = inputs["attention_mask"].unsqueeze(-1).float()
return (out.last_hidden_state * mask).sum(1) / mask.sum(1).clamp(min=1e-9)
```

Ein fester Seed allein reicht **nicht** — er macht das Ergebnis reproduzierbar, aber nicht sinnvoll.

### Eingriffspunkt

Der Defekt liegt in OASIS (`site-packages/oasis/social_platform/process_recsys_posts.py`), nicht in Agora-Code. Agora patcht dieses Modul aber bereits: `backend/scripts/_sim_common.py` installiert ab Zeile ~697 ein Ladeprofil für genau dieses Modell (`install_bert_memory_profile`, Kommentar ab Z. 680) und nutzt dafür Pythons Modul-Caching, damit der spätere OASIS-Import die gepatchte Methode sieht. Ein Monkey-Patch auf `process_batch` ließe sich am selben Ort und nach demselben Muster einhängen.

Alternativ upstream melden und bis dahin patchen.

## Nebenbefund: unauthentifizierter HF-Download bei jedem Start

Im selben Block erscheint:

```
huggingface_hub.utils._http - WARNING - You are sending unauthenticated requests to the HF Hub.
Please set a HF_TOKEN to enable higher rate limits and faster downloads.
```

Das Modell (1,06 GB) wird bei jedem Simulationsstart über das Netz aufgelöst. Für eine Plattform, die als „lokal oder kontrolliert hybrid betreibbar" beschrieben ist, ist das eine unbeabsichtigte externe Abhängigkeit im Startpfad — und ohne Token rate-limitiert.

## Akzeptanzkriterien

- [ ] Die Twitter-Post-Vektoren stammen aus trainierten Gewichten; kein `MISSING`-Key wird produktiv verwendet
- [ ] Zwei Läufe mit identischer Simulation und identischem Seed erzeugen identische `rec`-Matrizen
- [ ] Regressionstest, der `process_batch` gegen zwei semantisch nahe und zwei semantisch ferne Texte prüft und die Ähnlichkeitsordnung assertiert — gegen den aktuellen Stand instabil

## Beobachtet in

Läufe `sim_54c1c2a6a875` und `sim_8495a5fe314b` (11.08.2026), beide `runner_status: completed`, jeweils 10/10 Runden.

**Comment by arn0ld87** [comment]
> *This was generated by AI during triage.*

## Agent Brief

**Status:** `ready-for-agent`

**Verifikation:** Verifiziert: _sim_common.py patcht nur das TWHIN-BERT-Ladeprofil; es gibt keinen Agora-Patch für process_batch/pooler_output. Der beschriebene Random-Pooler-Pfad ist daher weiterhin offen. Auftrag: produktiven Twitter-Recommender auf trainierte/semantisch belastbare Repräsentation umstellen und deterministische Regression ergänzen.

**Redundanzprüfung:** Der aktuelle `main` wurde nach dem beschriebenen Mechanismus bzw. einem bereits vorhandenen äquivalenten Fix durchsucht. Wo Teilfixes existieren, sind sie im aktuellen Issue-Body ausdrücklich abgegrenzt.

**Arbeitsregel:** Der aktuelle Issue-Body ist die verbindliche Scope-/Akzeptanzquelle. Historische Kommentare nicht als Auftrag für bereits erledigte Teilprobleme behandeln.

**Definition of Done:** Alle verbleibenden Akzeptanzkriterien des aktuellen Bodys erfüllt, Regressionstest auf dem produktiven Codepfad, passende Gates grün, Changelog/STATUS gemäß AGENTS.md synchronisiert.

**Comment by arn0ld87** [comment]
Erledigt durch #1549 (1f2d3e67c): Twitter-Recommender rankt über Mean-Pooling statt über die zufallsinitialisierten Pooler-Gewichte von twhin-bert, inkl. Determinismus-Test und dtype-folgender Maske.

https://claude.ai/code/session_01TKQZSZg38jzFqY3zzXm2dX
